src/components/views/dialogs/CreateGroupDialog.js

```diff
@@ -0,0 +1,196 @@
+/*
+Copyright 2017 Vector Creations Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import sdk from '../../../index';
+import dis from '../../../dispatcher';
+import { _t } from '../../../languageHandler';
+import MatrixClientPeg from '../../../MatrixClientPeg';
+
+// We match fairly liberally and leave it up to the server to reject if
+// there are invalid characters etc.
+const GROUP_REGEX = /^\+(.*?):(.*)$/;
+
+export default React.createClass({
+    displayName: 'CreateGroupDialog',
+    propTypes: {
+        onFinished: PropTypes.func.isRequired,
+    },
+
+    getInitialState: function() {
+        return {
+            groupName: '',
+            groupId: '',
+            groupError: null,
+            creating: false,
+            createError: null,
+        };
+    },
+
+    _onGroupNameChange: function(e) {
+        this.setState({
+            groupName: e.target.value,
+        });
+    },
+
+    _onGroupIdChange: function(e) {
+        this.setState({
+            groupId: e.target.value,
+        });
+    },
+
+    _onGroupIdBlur: function(e) {
+        this._checkGroupId();
+    },
+
+    _checkGroupId: function(e) {
+        const parsedGroupId = this._parseGroupId(this.state.groupId);
+        let error = null;
+        if (parsedGroupId === null) {
+            error = _t(
+                "Group IDs must be of the form +localpart:%(domain)s",
+                {domain: MatrixClientPeg.get().getDomain()},
+            );
+        } else {
+            const domain = parsedGroupId[1];
+            if (domain !== MatrixClientPeg.get().getDomain()) {
+                error = _t(
+                    "It is currently only possible to create groups on your own home server: "+
+                    "use a group ID ending with %(domain)s",
+                    {domain: MatrixClientPeg.get().getDomain()},
+                );
+            }
+        }
+        this.setState({
+            groupIdError: error,
+        });
+        return error;
+    },
+
+    _onFormSubmit: function(e) {
+        e.preventDefault();
+
+        if (this._checkGroupId()) return;
+
+        const parsedGroupId = this._parseGroupId(this.state.groupId);
+        const profile = {};
+        if (this.state.groupName !== '') {
+            profile.name = this.state.groupName;
+        }
+        this.setState({creating: true});
```

Comment:
Same old "this might race". We should probably use the callback from `this.setState`. 

Looking into the whole "WHYNOGIVEPROMISE" problem of `setState`, it appears that the React team don't want it to return a promise and we shouldn't be using the callback anyway? https://github.com/facebook/react/pull/9989#issuecomment-309141521